consensus/consensus_v2.go

```diff
@@ -3,122 +3,651 @@ package consensus
 import (
 	"time"
 
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/rlp"
+
 	protobuf "github.com/golang/protobuf/proto"
+	"github.com/harmony-one/bls/ffi/go/bls"
+	"github.com/harmony-one/harmony/api/proto"
 	msg_pb "github.com/harmony-one/harmony/api/proto/message"
+	"github.com/harmony-one/harmony/api/service/explorer"
+	"github.com/harmony-one/harmony/core"
+	"github.com/harmony-one/harmony/core/types"
+	bls_cosi "github.com/harmony-one/harmony/crypto/bls"
 	"github.com/harmony-one/harmony/internal/utils"
+	"github.com/harmony-one/harmony/p2p"
+	"github.com/harmony-one/harmony/p2p/host"
 )
 
-// Start is the entry point and main loop for consensus
-func (consensus *Consensus) Start(stopChan chan struct{}, stoppedChan chan struct{}) {
-	defer close(stoppedChan)
-	tick := time.NewTicker(blockDuration)
-	consensus.idleTimeout.Start()
-	for {
-		select {
-		default:
-			msg := consensus.recvWithTimeout(receiveTimeout)
-			consensus.handleMessageUpdate(msg)
-			if consensus.idleTimeout.CheckExpire() {
-				consensus.startViewChange(consensus.consensusID + 1)
-			}
-			if consensus.commitTimeout.CheckExpire() {
-				consensus.startViewChange(consensus.consensusID + 1)
-			}
-			if consensus.viewChangeTimeout.CheckExpire() {
-				if consensus.mode.Mode() == Normal {
-					continue
-				}
-				consensusID := consensus.mode.ConsensusID()
-				consensus.startViewChange(consensusID + 1)
-			}
-		case <-tick.C:
-			consensus.tryPublishBlock()
-		case <-stopChan:
-			return
-		}
-	}
-
-}
-
-// recvWithTimeout receives message before timeout
-func (consensus *Consensus) recvWithTimeout(timeoutDuration time.Duration) []byte {
-	var msg []byte
-	select {
-	case msg = <-consensus.MsgChan:
-		utils.GetLogInstance().Debug("[Consensus] recvWithTimeout received", "msg", msg)
-	case <-time.After(timeoutDuration):
-		utils.GetLogInstance().Debug("[Consensus] recvWithTimeout timeout", "duration", timeoutDuration)
-	}
-	return msg
-}
-
 // handleMessageUpdate will update the consensus state according to received message
 func (consensus *Consensus) handleMessageUpdate(payload []byte) {
+	if len(payload) == 0 {
+		return
+	}
 	msg := &msg_pb.Message{}
 	err := protobuf.Unmarshal(payload, msg)
 	if err != nil {
 		utils.GetLogInstance().Error("Failed to unmarshal message payload.", "err", err, "consensus", consensus)
+		return
+	}
+
+	consensusMsg := msg.GetConsensus()
+	senderKey, err := bls_cosi.BytesToBlsPublicKey(consensusMsg.SenderPubkey)
+	if err != nil {
+		utils.GetLogInstance().Debug("Failed to deserialize BLS public key", "error", err)
+		return
+	}
+	addrBytes := senderKey.GetAddress()
+	senderAddr := common.BytesToAddress(addrBytes[:])
+
+	if !consensus.IsValidatorInCommittee(senderAddr) {
+		utils.GetLogInstance().Error("Invalid validator", "senderAddress", senderAddr)
+		return
+	}
+
+	if err = verifyMessageSig(senderKey, msg); err != nil {
```

Comment:
I think we should first check the message type before verifying signature. Sig verification is expensive.

Reply:
If we check the type and parse it in different handler, we still need to change the signature, since there are not other type of messages other than in the switch condition.